### PR TITLE
fix for #1363 - use clearMap instead of reset to clear costmaps

### PR DESCRIPTION
**Pull request by mlherd**
- This PR is a fix for #1363. 
- Created clearMap function for each layer
- Use clearMap() instead of reset() to clear the costmaps because reset is costly since it does so many unnecessary stuff

### CODE REVIEW

File: nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp [hunk 1]
```diff
@@ -70,6 +70,7 @@ class Layer  // TODO(mjeronimo): public nav2_util::LifecycleHelperInterface
   virtual void deactivate() {} /** @brief Stop publishers. */
   virtual void activate() {}   /** @brief Restart publishers if they've been stopped. */
   virtual void reset() {}
+  virtual void clearMap() {}
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp]: This needs to be pure virtual. It must be implemented

File: nav2_costmap_2d/plugins/obstacle_layer.cpp [hunk 2]
```diff
@@ -658,4 +653,22 @@ ObstacleLayer::reset()
   undeclareAllParameters();
 }
 
+void
+ObstacleLayer::triggerObservationsBufferUpdate()
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/obstacle_layer.cpp]: I’m not sure this method name describes what’s happening here. Given its doing a single operation over all streams, how about `resetBuffersLastUpdated`?

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 3]
```diff
@@ -169,15 +169,15 @@ void VoxelLayer::reset()
...
+  clearMap();
   activate();
   undeclareAllParameters();
   voxel_pub_->on_activate();
 }
 
-void VoxelLayer::resetMaps()
+void VoxelLayer::clearMap()
 {
-  Costmap2D::resetMaps();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp]: As far as I can tell from this diff, you didnt remove resetMaps from costmap2D. Should you? What’s the difference between clear and reset?

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: You also called the costmap2D directly not the obstacle layer this derives from that also needs to do useful “stuff”

**Comment by SteveMacenski** [comment]
I think this is going to start getting more involved than you bargained for. Why not remove the unsubscribe/resubscribe & replace with a flag in the callbacks to ignore incoming information or something? That seems to accomplish the same thing but without changing the structure of clear/resets

File: nav2_costmap_2d/plugins/voxel_layer.cpp (hunk 3 above)
Comment (mlherd) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: reset - calls resetMaps, initializes everything including parameters, subscribers and publishers
clearMap - calls resetMaps or/and sets the necessary flags

Comment (mlherd) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: I will test this to simplify clearMap in voxelLayer.

VoxelLayer::clearMap()
{
  resetMaps();
  ...
}

File: nav2_costmap_2d/plugins/obstacle_layer.cpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/obstacle_layer.cpp]: (its also just iterating through a vector, I dont think we need to over complicate it, it could also just be copy pasted twice and I wouldnt argue)

File: nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp [hunk 4]
```diff
@@ -80,9 +80,7 @@ class VoxelLayer : public ObstacleLayer
   }
   virtual void matchSize();
   virtual void reset();
-
-protected:
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp]: https://github.com/ros-planning/navigation2/blob/4f42633b9320d503e0f26d1fc7773485b4f94de3/nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp#L362

File: nav2_costmap_2d/plugins/voxel_layer.cpp (hunk 3 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: resetMaps is implemented all over the place. I really dont think you should be messing with this. 

As far as I can tell, the only think you really need to do is change the deactivate() and activate() methods to not unsubscribe from sensor readings (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L634) and then register the callbacks with their `is_active_[subscriber]` bool (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L207). Callback will be called but then the value is immediately passed on in the callback using that bool.

Right now you don't actually stop data from coming in and being processed on reset. This has a number of unintended consequences.

Comment (mlherd) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: I don't think it is that complicated. resetMaps is implemented in costmap2D. In costmap2D it used is when initializing and resizing cost maps. Other than that only the obstacle layer and voxel layer uses to clear their costmaps.

All it does is setting the costmap values to some default values.

```
void Costmap2D::resetMaps()
{
  std::unique_lock<mutex_t> lock(*access_);
  memset(costmap_, default_value_, size_x_ * size_y_ * sizeof(unsigned char));
}
```

--- 

> This has a number of unintended consequences.

Could you give an example? I am just trying to understand how to solve fix this issue without introducing new issues.


>  As far as I can tell, the only thing you really need to do is change the deactivate() and activate() methods to not unsubscribe from sensor readings (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L634) and then register the callbacks with their is_active_[subscriber] bool (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L207). Callback will be called but then the value is immediately passed on in the callback using that bool.

I will give this solution a try tomorrow.

Comment (crdelsey) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: > resetMaps is implemented all over the place. I really dont think you should be messing with this (deleting in some places, but not others, reset doesnt seem aligned with all the plugins...).

resetMaps is part of the costmap2d class interface, not the layer interface. I expect that is why it doesn't line up exactly with the plugins. I'm not sure all the plugins inherit from both layer and costmap2d.

Every plugin needs a reset and a clearMap method since that's part of the layer interface. During those methods they may need to call resetMaps if they depend on an underlying costmap2d object.

> As far as I can tell, the only thing you really need to do is change the deactivate() and activate() methods to not unsubscribe from sensor readings

I believe activate and deactive are used by the lifecycle transitions. They need to do subscribing and unsubscribing to go through the lifecycle stages. I think it is better to just remove them from the clear costmap flow, which, I believe is what @mlherd is doing with this PR.

> Right now you don't actually stop data from coming in and being processed on reset. This has a number of unintended consequences.

If I understand your concern correctly, I believe this is safe. There are two threads in costmap, the map update thread and the main spin loop. The clearEntireCostmap service is called on the main spin loop, so no other messages can get processed until it completes. The update loop is block by a mutex taken during the processing of the service request, before the individual plugins are called to clear.

**Comment by crdelsey** [comment]
@SteveMacenski  Are you ok with these changes now?

**Comment by SteveMacenski** [comment]
No, I really do not think this is the right way to go, I think it adds complexity and ambiguity. 

This issue you're trying to solve is not toggling the subscribers, that should be what is directly solved

**Comment by crdelsey** [comment]
>  that should be what is directly solved by replacing that behavior using the enabled characteristics

I don't understand.

I don't see a simpler way to do this other than deleting the old reset function and replacing with these new clearMap functions.

Basically, @mlherd has taken the old function and deleted out the subscribe/unsubscribe and fixed a few bugs in the process.

**Comment by SteveMacenski** [comment]
Lets follow the trail of how it worked before, and how its changed.

In costmap 2d ROS we'd have a `plugin->reset()` method. `reset()` was a virtual method with a default implementation `{}` in the header. 

Then if we look into a specific layer, lets say the static layer, the reset method resets the subscribers undeclares and sets up the plugin again. 

Now in this PR, we replace `reset()` with `clearMap()` as our entry point for costmap 2d ROS, similarly with a virtual method with default implementation`{}` in the cpp file. 

In the static layer, the `clearMap()` method will set some parameter, not sure why its here and not in `reset()` but not really the point.

Ok. So now we have this `reset()` method **that was never removed** without anyone calling it, and then `clearMap()`. That's the first issue I do not like, we're renaming methods and swapping them out, and then not removing the initial implementations. At that point, why make a new method? That doesn't serve any value other than to confuse, we should be changing the implementations of `reset()`, not replacing the method entirely and then not removing the original.

In the voxel layer, we're renaming `resetMaps` to `clearMaps`, I don't know why, but again, less the point. Within the new `clearMaps`, `Costmap2D::resetMaps();` was replaced with `ObstacleLayer::clearMap();`, which does entirely different things. Within `ObstacleLayer::clearMap();`, we will eventually call `Costmap2D::resetMaps();`, but now we're also resetting the buffers which wasn't happening before in the voxel layer. I don't know what the performance or other implications of that are, but seems to be a hidden result of that choice that I don't understand off hand how it will affect things.

Finally, the `reset()`, `resetMap()`, `resetMaps()` is a consistent naming nomenclature across the base class `Costmap2D`, the costmap plugin layers, the higher level `LayeredCostmap`, and the application layer `Costmap2DROS` to do specific and consistent things. Renaming this could potentially break inheritance properties and also breaks some naming conventions. 

tl;dr
I have no issue with what this is trying to do, but I see renaming methods as a detriment, leaving old implementations around but unused is confusing, and added behaviors that I dont think have been fully thought about. This is why I recommend only changing the method `reset{s, Map, Maps}` implementations and not adding some additional methods.

**Comment by crdelsey** [comment]
>  That doesn't serve any value other than to confuse, we should be changing the implementations of reset(), not replacing the method entirely and then not removing the original.

I agree. We should delete the old methods.

> In the voxel layer, we're renaming resetMaps to clearMaps, I don't know why, but again, less the point. Within the new clearMaps, Costmap2D::resetMaps(); was replaced with ObstacleLayer::clearMap();, which does entirely different things. Within ObstacleLayer::clearMap();, we will eventually call Costmap2D::resetMaps();, but now we're also resetting the buffers which wasn't happening before in the voxel layer.

This is the part that is fixing some bugs. The old voxel layer was not resetting things cleanly. It bypassed the obstacle layer it was based on, preventing it from properly cleaning up. I believe the new behavior here is much more correct.

----

I'd like to keep the new voxel layer implementation. I believe the old one was broken.

I'm hearing we need to delete the old reset functions and rename the new clearMap functions as reset. This preserves the old interface.

This will have the effect that reset only clears data structures, triggers a refresh of the costmap layer and no longer mucks with the topic subscription.

**Comment by SteveMacenski** [comment]
> I'm hearing we need to delete the old reset functions and rename the new clearMap functions as reset. This preserves the old interface.

Yes, just change the things happening in the existing methods.

> This is the part that is fixing some bugs. The old voxel layer was not resetting things cleanly. It bypassed the obstacle layer it was based on, preventing it from properly cleaning up. I believe the new behavior here is much more correct.

I don't disagree it may be a benefit, but that may have been done on purpose in ROS1, I really just don't know. I think that needs to be stress tested.

> This will have the effect that reset only clears data structures, triggers a refresh of the costmap layer and no longer mucks with the topic subscription.

Yes, but should be stress tested with clearing a bunch to make sure it doesnt have some weird artifacts happening

**Comment by crdelsey** [comment]
> Yes, but should be stress tested with clearing a bunch to make sure it doesnt have some weird artifacts happening

@mlherd We tested this didn't we? Before the changes, the voxel layer wasn't getting cleared properly. Afterwards it worked, as I recall. I might be getting mixed up with obstacle layer though. Can you confirm?

**Comment by crdelsey** [comment]
FYI: Making the change to alter the existing reset method instead of creating a new one will conflict with #1431. Be careful during the merge.

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 5]
```diff
@@ -167,17 +167,7 @@ void VoxelLayer::matchSize()
...
 {
-  voxel_pub_->on_deactivate();
-  deactivate();
-  resetMaps();
-  activate();
-  undeclareAllParameters();
-  voxel_pub_->on_activate();
-}
-
-void VoxelLayer::resetMaps()
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp]: Where did the `resetMaps` method in Voxel layer go? It looks like you replaced it work only `reset()`, why?

Comment (crdelsey) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: The relevant inheritance hierarchy looks like this:

image

Voxel layer was overriding the Costmap2D method only to call it directly without adding any functionality. It served no purpose.

What's more, if obstacle layer at sometime decided to override resetMaps, VoxelLayer would have bypassed it.

**Comment by crdelsey** [comment]
Since you've had a chance to look at the changes, I'm going to rebase the branch to get it to pass CI.

File: nav2_costmap_2d/plugins/voxel_layer.cpp (hunk 5 above)
Comment (crdelsey) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: Let me add, resetMaps is the method to reset the Costmap2D base class and reset is the method to reset the Layer base class.

They are two separate methods in the interface. We now override reset in VoxelLayer and no longer override resetMaps.

It's not a matter of changing the method name.

image

Comment (crdelsey) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: This maybe is a slightly better diagram
image

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: That's untrue, see here https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/voxel_layer.cpp#L121. The reset method is what is actually clearing the voxel grid data structure. 

`resetMaps` is then called by the costmap 2d layer in the rolling costmap mode which will clear out the rolling window of voxels as well for the current set.
